src/Sources/Coalescence.jl

```diff
@@ -3,425 +3,524 @@
     end
 end
 
-function get_coalescence_integral_moment_qrs!(
+function get_coalescence_integral_moment_qrs(
     cs::AnalyticalCoalStyle,
-    moment_order::Int,
-    NProgMoms::Vector{Int},
-    coal_data,
-)
-    order = size(coal_data.matrix_of_kernels[1].c)[1] - 1
-    update_Q_coalescence_matrix!(
-        cs,
-        moment_order,
-        coal_data.moments,
-        order,
-        coal_data.matrix_of_kernels,
-        NProgMoms,
-        coal_data.Q,
-    )
-    update_R_coalescence_matrix!(
-        cs,
-        moment_order,
-        coal_data.moments,
-        order,
-        coal_data.matrix_of_kernels,
-        NProgMoms,
-        coal_data.R,
-    )
-    update_S_coalescence_matrix!(
-        cs,
-        moment_order,
-        coal_data.moments,
-        order,
-        coal_data.finite_2d_ints,
-        coal_data.matrix_of_kernels,
-        NProgMoms,
-        coal_data.S,
+    moments::SMatrix{N, M, FT},
+    NProgMoms::NTuple{N, Int},
+    finite_2d_ints::NTuple{N, SMatrix},
+    kernels::NTuple{N, NTuple{N, CoalescenceTensor{P, FT}}},
+) where {N, M, P, FT <: Real}
+    return (;
+        Q = get_Q_coalescence_matrix(cs, moments, NProgMoms, kernels),
+        R = get_R_coalescence_matrix(cs, moments, NProgMoms, kernels),
+        S = get_S_coalescence_matrix(cs, moments, NProgMoms, finite_2d_ints, kernels),
     )
 end
 
-function update_Q_coalescence_matrix!(
-    ::AnalyticalCoalStyle,
-    moment_order,
-    moments,
-    order,
-    matrix_of_kernels,
-    NProgMoms,
-    Q,
-)
-    Ndist = size(moments)[1]
-
-    for j in 1:Ndist
-        for k in (j + 1):Ndist
-            Q[j, k] = 0.0
-            if NProgMoms[k] <= moment_order
-                continue
+function get_Q_coalescence_matrix(
+    cs::AnalyticalCoalStyle,
+    moments::SMatrix{N, M, FT},
+    NProgMoms::NTuple{N, Int},
+    kernels::NTuple{N, NTuple{N, CoalescenceTensor{P, FT}}},
+) where {N, M, P, FT <: Real}
+
+    return ntuple(3) do i
+        moment_order = i - 1
+        SMatrix{N, N, FT}(rflatten(ntuple(N) do k
+            ntuple(N) do j
+                if k <= j || NProgMoms[k] <= moment_order
+                    FT(0)
+                else
+                    Q_jk(cs, moment_order, j, k, moments, kernels[j][k])
+                end
             end
-            r = order
-            for a in 0:r
-                for b in 0:r
-                    for c in 0:moment_order
-                        Q[j, k] +=
-                            matrix_of_kernels[j, k].c[a + 1, b + 1] *
+        end))
+    end
+end
+
+function Q_jk(
+    ::AnalyticalCoalStyle,
+    moment_order::Int,
+    j::Int,
+    k::Int,
+    moments::SMatrix{N, M, FT},
+    kernel::CoalescenceTensor{P, FT},
+) where {N, M, P, FT <: Real}
+    return sum(
+        ntuple(P) do a1
+            a = a1 - 1
+            sum(
+                ntuple(P) do b1
+                    b = b1 - 1
+                    sum(
+                        map(1:(moment_order + 1)) do c1
```

Comment:
Do we want to use a map over an ntuple here?

Reply:
`map` is fine, but we could just use `sum(c1->f(c1), 1:(moment_order + 1))` syntax instead.